File: src/core/scala/mrtjp/core/vec/Point.scala
/*
 * Copyright (c) 2014.
 * Created by MrTJP.
 * All rights reserved.
 */
package mrtjp.core.vec

case class Point(x:Int, y:Int)
{
    override def equals(obj:scala.Any) = obj match
    {
        case that:Point => x == that.x && y == that.y
        case _ => false
    }

    def copy = Point(x, y)

    def add(dx:Int, dy:Int) = Point(x+dx, y+dy)
    def subtract(dx:Int, dy:Int) = add(-dx, -dy)
    def multiply(i:Int, j:Int) = Point(x*i, y*j)
    def divide(i:Int, j:Int) = Point(x/i, y/j)

    def add(d:Int):Point = add(d, d)
    def subtract(d:Int):Point = subtract(d, d)
    def multiply(k:Int):Point = multiply(k, k)
    def divide(k:Int):Point = divide(k, k)

    def add(that:Point):Point = add(that.x, that.y)
    def subtract(that:Point):Point = subtract(that.x, that.y)
    def multiply(that:Point):Point = multiply(that.x, that.y)
    def divide(that:Point):Point = divide(that.x, that.y)

    def negate = Point(-x, -y)
    def invert = Point(y, x)
    def vectorize = Vec2(x, y)

    def max(that:Point) = Point(x max that.x, y max that.y)
    def min(that:Point) = Point(x min that.x, y min that.y)

    def clamp(rect:Rect) = this min rect.maxPoint max rect.origin
    def clamp(size:Size) = this min Point(size) max Point.zeroPoint

    def offset(r:Int):Point = offset(r, 1)
    def offset(r:Int, amount:Int):Point = this+(Point.dirOffsets(r)*amount)

    def unary_- = negate
    def unary_~ = invert

    def +(that:Point) = add(that)
    def -(that:Point) = subtract(that)
    def *(that:Point) = multiply(that)
    def /(that:Point) = divide(that)

    def +(that:Int) = add(that)
    def -(that:Int) = subtract(that)
    def *(that:Int) = multiply(that)
    def /(that:Int) = divide(that)

    override def toString = s"Point @[$x $y]"
}

object Point
{
    val infinitePoint = Point(Int.MaxValue, Int.MaxValue)
    val zeroPoint = Point(0, 0)

    def apply(size:Size):Point = Point(size.width, size.height)

    def apply(vec2:Vec2):Point = Point(vec2.dx.toInt, vec2.dy.toInt)

    private val dirOffsets = Seq(
        Point(0, -1),
        Point(1, 0),
        Point(0, 1),
        Point(-1, 0)
    )
}

File: src/core/java/mrtjp/projectred/lib/Point.java
package mrtjp.projectred.lib;

public class Point {

    public static final Point ZERO = new Point(0, 0);
    public static final Point INFINITE = new Point(Integer.MAX_VALUE, Integer.MAX_VALUE);

    public static final Point[] DIR_OFFSETS = { new Point(0, -1), new Point(1, 0), new Point(0, 1), new Point(-1, 0) };

    public final int x;
    public final int y;

    public Point(int x, int y) {
        this.x = x;
        this.y = y;
    }

    public Point add(Point p) {
        return new Point(x + p.x, y + p.y);
    }

    public Point subtract(Point p) {
        return new Point(x - p.x, y - p.y);
    }

    public Point multiply(Point p) {
        return new Point(x * p.x, y * p.y);
    }

    public Point divide(Point p) {
        return new Point(x / p.x, y / p.y);
    }

    public Point add(int x, int y) {
        return new Point(this.x + x, this.y + y);
    }

    public Point subtract(int x, int y) {
        return new Point(this.x - x, this.y - y);
    }

    public Point multiply(int x, int y) {
        return new Point(this.x * x, this.y * y);
    }

    public Point divide(int x, int y) {
        return new Point(this.x / x, this.y / y);
    }

    public Point negate() {
        return new Point(-x, -y);
    }

    @SuppressWarnings("SuspiciousNameCombination")
    public Point flip() {
        return new Point(y, x);
    }

    public Vec2 toVec2() {
        return new Vec2(x, y);
    }

    public Point max(Point p) {
        return new Point(Math.max(x, p.x), Math.max(y, p.y));
    }

    public Point min(Point p) {
        return new Point(Math.min(x, p.x), Math.min(y, p.y));
    }

    public Point max(int x, int y) {
        return new Point(Math.max(this.x, x), Math.max(this.y, y));
    }

    public Point min(int x, int y) {
        return new Point(Math.min(this.x, x), Math.min(this.y, y));
    }

    public Point clamp(Rect rect) {
        return this.min(rect.maxPoint()).max(rect.origin);
    }

    public Point clamp(Size size) {
        return this.min(size.width, size.height).max(ZERO);
    }

    public Point offset(int r) {
        return offset(r, 1);
    }

    public Point offset(int r, int amount) {
        return new Point(x + DIR_OFFSETS[r].x * amount, y + DIR_OFFSETS[r].y * amount);
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof Point) {
            Point p = (Point) obj;
            return x == p.x && y == p.y;
        }
        return false;
    }

    @Override
    public String toString() {
        return "Point[" + x + ", " + y + "]";
    }
}
